File: samples-java/src/main/java/io/github/ilaborie/loops/samples/recursion/AccumulateTR.java
package io.github.ilaborie.loops.samples.recursion;

import io.github.ilaborie.loops.samples.Element;

import java.util.List;

import static io.github.ilaborie.loops.samples.Elements.getSomeElements;

@SuppressWarnings("ALL")
public class AccumulateTR {

    public static void main(String[] args) {
        List<Element> elements = getSomeElements();
        int result = accumulate(elements);
        System.out.println(result);
    }

    public static int accumulate(List<Element> elements) {
        return aux(elements, 0);
    }

    public static int aux(List<Element> elements, int acc) {
        if (elements.isEmpty()) {
            return acc;
        }

        Element head = elements.get(0);
        List<Element> tail = elements.subList(1, elements.size());

        int count = head.getName().length();
        return aux(tail, acc + count);
    }

}

File: samples-kotlin/src/main/kotlin/io/github/ilaborie/loops/samples/recursion/accumulateTR.kt
package io.github.ilaborie.loops.samples.recursion

import io.github.ilaborie.loops.samples.ElementKt
import io.github.ilaborie.loops.samples.getSomeElementsKt


fun main() {
    val elements = getSomeElementsKt()
    val result = accumulateTR(elements)
    println(result)
}

fun accumulateTR(elements: List<ElementKt>): Int {
    tailrec fun aux(elements: List<ElementKt>, acc: Int): Int {
        if (elements.isEmpty()) {
            return acc
        }

        val head = elements.first()
        val tail = elements.drop(1)

        val count = head.name.length
        return aux(tail, acc + count)
    }
    return aux(elements, 0)
}

fun accumulateTRAlt(elements: List<ElementKt>): Int {
    tailrec fun aux(elements: List<ElementKt>, acc: Int): Int =
        if (elements.isEmpty()) acc
        else aux(elements.drop(1), acc + elements.first().name.length)

    return aux(elements, 0)
}
